### ISSUE TITLE
C++/Python with shared_ptr lists: Assertion `own' failed.

### ISSUE THREAD

**Issue by nschloe** [swig/swig#773]
In a SWIGged C++/Python application, I'd like to create a function that accepts a `std::vector`/Python list of abstract base class objects, e.g.,

``` c++
#ifndef MYTEST_HPP
#define MYTEST_HPP

#include <iostream>
#include <vector>
#include <memory>

class MyBaseClass {
  public:
  virtual void eval() const = 0;
};

class MyTestClass: public MyBaseClass {
  public:
  virtual void eval() const {
    std::cout << "MyTestClass::eval()" << std::endl;
  }
};

void
mytest(const std::vector<std::shared_ptr<MyBaseClass>> & a) {
  std::cout << "length: " << a.size() << std::endl;
}

#endif // MYTEST_HPP
```

Note that we're using `std::shared_ptr`s here. Using this with

``` swig
%module mytest

%{
#define SWIG_FILE_WITH_INIT
#include "mytest.hpp"
%}

%include <std_shared_ptr.i>
%shared_ptr(MyBaseClass);
%shared_ptr(MyTestClass);

%include <std_vector.i>
namespace std {
%template(List) vector<std::shared_ptr<MyBaseClass>>;
}

%include "mytest.hpp"
```

does compile, but fails at runtime. From

``` python
import mytest
a = mytest.MyTestClass()
mytest.mytest([a])
```

I'm getting

```
python: /tmp/l/build/mytestPYTHON_wrap.cxx:2385: int SWIG_Python_ConvertPtrAndOwn(PyObject*, void**, swig_type_info*, int, int*): Assertion `own' failed.
Aborted (core dumped)
```

[This works with SWIG 3.0.2](https://stackoverflow.com/questions/39163006/vector-list-of-abstract-base-class-objects-in-swig-c-python), but fails at least since 3.0.8.

Note that with

``` python
import mytest
a = mytest.MyTestClass()
l = mytest.List()
l.append(a)
mytest.mytest(l)
```

works.

---

_Note:_ The corresponding Gist is [here](https://gist.github.com/nschloe/c6d1dcb0837031183da0da5423ea23f6).

**Comment by nschloe** [comment]
The same happens with [the latest SWIG from master](https://launchpad.net/~nschloe/+archive/ubuntu/swig-nightly/).

**Comment by wsfulton** [comment]
Both 3.0.2 and latest master assert. Neither versions assert if you compile with -DNDEBUG. I'm not too surprised this doesn't work as I think both std_vector.i and shared_ptr.i both define custom typemaps and possibly the same ones. I'll take a look when I get a chance.

**Comment by q-p** [comment]
Possible dup of (or same cause as)  #731

**Comment by autosquid** [comment]
any progress on this?

**Comment by ddurham2** [comment]
@wsfulton, I'd really like to see this problem fixed.  Did you ever have a chance to look into it?  I'm willing to pay you money and/or beer. :)


FWIW, I've set up a repo that demonstrates the same: https://github.com/ddurham2/swig-prob1

You are correct that -DNDEBUG will suppress the assert, but (as mine demonstrates) it does leak an object (yay assert!) which is a major problem for me.

Thanks!

**Comment by ddurham2** [comment]
So, I did some more digging.  If you'll permit me to have a little of your time for the sake of improving swig, I'll try to be as succinct as possible but with helpful detail...

In short, the problem here (#773) is that a leak occurs while constructing a swig object (in python) which wraps a vector<shared_ptr<T>>.  Constructing one from a python list[] leaks the objects in the list presumably because the given shared_ptr<T>'s get new-copy-constructed-ed, but never delete-ed.

Let our T type be Foo.  And also Bar derives from Foo.

The problem is actually two-fold.  _wrap_new_vector_Foo() has two phases.  The code looks like this:

    ... 
      if (argc == 1) {
        int _v;
        int res = swig::asptr(argv[0], (vector<shared_ptr<Foo> >**)(0));      <------ (*1*) checking before creating
        _v = SWIG_CheckState(res);
        if (_v) {
          return _wrap_new_vector_Foo__SWIG_1(self, args);   <------ (*2*) creating
        }
      }
    ...


First it calls swig::asptr() with a null output parameter which apparently indicates that it should just check if the type conversion is okay and not actually return a created object.

... and the call stack looks like this (simplified for brevity):

    assert(own)
    SWIG_Python_ConvertPtrAndOwn (obj=0x7fffe2bd6710, ptr=0x7fffffffce00, ty=0x7ffff662c140 <_swigt__p_std__shared_ptrT_Foo_t>, flags=0, own=0x0)
    swig::traits_asptr<shared_ptr<Foo> >::asptr (obj=0x7fffe2bd6710, val=0x0)
    swig::asptr<shared_ptr<Foo> > (obj=0x7fffe2bd6710, vptr=0x0)
    swig::traits_check<shared_ptr<Foo>, swig::pointer_category>::check (obj=0x7fffe2bd6710)
    swig::check<shared_ptr<Foo> > (obj=0x7fffe2bd6710)
    swig::SwigPySequence_Cont<shared_ptr<Foo> >::check (this=0x7fffffffd300, set_err=true)
    swig::traits_asptr_stdseq<vector<shared_ptr<Foo> >, shared_ptr<Foo> >::asptr (obj=0x7fffe305de88, seq=0x0)
    swig::traits_asptr<vector<shared_ptr<Foo> > >::asptr (obj=0x7fffe305de88, vec=0x0)
    swig::asptr<vector<shared_ptr<Foo> > > (obj=0x7fffe305de88, vptr=0x0)
    _wrap_new_vector_Foo (self=0x7ffff6b1cb38, args=0x7fffe2bd62b0)
        ... snip ...

It's in this phase that we hit the assert() in SWIG_Python_ConvertPtrAndOwn which actually has a comment about a memory leak.

    int SWIG_Python_ConvertPtrAndOwn(PyObject *obj, void **ptr, swig_type_info *ty, int flags, int *own) {
    ...
      while (sobj) {
        void *vptr = sobj->ptr;
        if (ty) {
          swig_type_info *to = sobj->ty;
          if (to == ty) {
            /* no type cast needed */
            if (ptr) *ptr = vptr;
            break;    <----- types are equivalent, so no work to do
          } else {
            swig_cast_info *tc = SWIG_TypeCheck(to->name,ty);
            if (!tc) {
              sobj = (SwigPyObject *)sobj->next;
            } else {
              if (own) {
                int newmemory = 0;
                *ptr = SWIG_TypeCast(tc,vptr,&newmemory);
                if (newmemory == SWIG_CAST_NEW_MEMORY) {
                  assert(own); /* badly formed typemap which will lead to a memory leak - it must set and use own to delete *ptr */   <-------
                  if (own)
                    *own = *own | SWIG_CAST_NEW_MEMORY;
                }
              }
              break;
            }
          }
        } else {
          if (ptr) *ptr = vptr;
          break;
        }
      }
    ...
    }


Within this SWIG_Python_ConvertPtrAndOwn() function, the from-type (ty) is shared_ptr<Foo> and the to-type (sobj->ty) is shared_ptr<Bar> (Bar is derived from Foo).  If the types were equivalent, then it would have no work to do and just return success.  However, since the types are different it has to check a derivation chain for compatibility (using SWIG_TypeCheck()).  In this case it does find a match and considers the types compatible.  Then it proceeds to call SWIG_TypeCast() which eventually calls "new shared_ptr<Bar>()" and sets newmemory to SWIG_CAST_NEW_MEMORY, but since own was passed in as null, it asserts.

I think the fix for this is that traits::asptr() should pass a null value for ptr when calling SWIG_ConvertPtrAndOwn(), if was passed a null value for val.  This will make it not even create the object that's just going to get thrown away anyway.  That is, unless you know why it's actually necessary to create that object.

I test this and it indeed resulted in the reference count of the leaked shared_ptr to be one less.

Okay, on to the second phase.

Here, a similar code path is followed where traits::asptr() is called, except now val is not null.  And it does not pass 'own' to SWIG_Python_ConvertPtrAndOwn() which causes it to assert too.

To fix this, it seems at this point traits::asptr() should (like other places I see in the ..._wrap.cpp file) declare an int newmem=0 and pass that for the own parameter to SWIG_Python_ConvertPtrAndOwn().  Then when that returns, we check newmem for the SWIG_CAST_NEW_MEMORY bit and subsequently set the SWIG_NEWOBJMASK on the return value so that the caller will know that it needs to free the memory.
---

I tried both of these fixes, and it seems to fix the problem in my test app.  For all this I was just hacking on the generated file, but I need to work on a PR that changes swig itself.

Please let me know if you think there are other places that be problematic for the same reason

**Comment by wsfulton** [comment]
Fixed, see #1512.
